### PR TITLE
MuReNNDirect

### LINKED ISSUE #34: `MuReNNDirect`
with parameters `J`, `Q`, `T`, `padding_mode`
and any other parameter you deem necessary from either DTCWT or Conv1D

good luck! 🍀

### PR DESCRIPTION
**Pull request by xir4n**
[EDIT by @lostanlen: fixes #34 ]

### CODE REVIEW

File: murenn/__init__.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 # Submodule imports
 from .dtcwt.transform1d import DTCWTDirect, DTCWTInverse
 from .version import version as __version__
+from .dtcwt.graph import MURENN_GRAPH
```
Comment (lostanlen) [inline murenn/__init__.py:17]: `from .dtcwt.murenn import MuReNNDirect`

File: murenn/dtcwt/graph.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import torch
+import murenn
+
+
+class MURENN_GRAPH(torch.nn.Module):
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:5]: `class MuReNNDirect`

File: murenn/dtcwt/graph.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    def __init__(
+        self,
+        J=8,
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:8]: no default values for keyword arguments `J`, `T`, `Q`

use required keyword arguments
`def __init__(self, *, J, Q, T, in_channels, padding_mode='symmetric')`

File: murenn/dtcwt/graph.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        Q=4,
+        T=16,
+        padding_mode="symmetric",
+    ):
+        super().__init__()
+        self.J = J
+        self.Q = Q
+        self.T = T
+        self.padding_mode = padding_mode
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:17]: no need to store padding mode and J twice. it's already in DTCWT ?

File: murenn/dtcwt/graph.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        self.dtcwt = murenn.DTCWT(
+            J=self.J,
+            padding_mode=self.padding_mode,
+        )
+
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:22]: initialize `Conv1D` and downsampling operators here, not in forward

File: murenn/dtcwt/graph.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    def forward(self, x):
+        _, bps = self.dtcwt(x)
+        ys = []
+        C = bps[0].shape[1]
+
+        for j in range(self.J):
+            down = murenn.DTCWT(
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:30]: move to `__init__`

File: murenn/dtcwt/graph.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                J=self.J-j,
+                padding_mode=self.padding_mode,
+                skip_hps=True,
+            )
+
+            conv1d = torch.nn.Conv1d(
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:36]: move to `__init__`

File: murenn/dtcwt/graph.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        self.dtcwt = murenn.DTCWT(
+            J=self.J,
+            padding_mode=self.padding_mode,
+        )
+
+
+    def forward(self, x):
+        _, bps = self.dtcwt(x)
+        ys = []
+        C = bps[0].shape[1]
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:27]: should be a keyword argument `in_channels`

File: murenn/dtcwt/graph.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                in_channels=C,
+                out_channels=self.Q*C,
+                kernel_size=self.T,
+                bias=False,
+                groups=C,
+                padding="same",
+            )
+            torch.nn.init.normal_(conv1d.weight) #std?
+
+            y_j, _ = down(conv1d(torch.abs(bps[j])))#B*(C*Q)*N
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:46]: should be three or four lines

File: murenn/dtcwt/graph.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            ys.append(y_j)
+
+        ys = torch.stack(ys, dim=2) #B*(C*Q)*J*N
```
Comment (lostanlen) [inline murenn/dtcwt/graph.py:49]: `# B * C * Q * J * N_J`

or

`# B * C * J * Q * N_J`

File: tests/test_graph.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    B, C, N = 2, 3, 2**(J+4)
+    x = torch.zeros(B, C, N)
+    graph = murenn.MURENN_GRAPH(
+        J=J,
+        Q=Q,
+        T=T,
+        padding_mode=padding_mode,
+    )
+    y = graph(x)
+    assert y.shape == (B, C*Q, J, 2**4)
```
Comment (lostanlen) [inline tests/test_graph.py:20]: update shape

Comment (lostanlen) [reply tests/test_graph.py:20]: test differentiability: require grad in the weights of the Conv1D's
define a loss function based on the output of MuReNNDirect. For example: the sum of all coefficients
Compute `w.grad`
It should be float32, not `None`

**Review by lostanlen — CHANGES_REQUESTED** [review]
thank you! good luck with future edits 🍀

**Review by lostanlen — APPROVED** [review]
Excellent work. I only have one remark about the name of the module. `graph.py` does not seem ideal. Maybe `nn.py` ? This wait it mirror `torch.nn`

File: murenn/__init__.py [hunk 12]
```diff
@@ -14,6 +14,7 @@
 # Submodule imports
 from .dtcwt.transform1d import DTCWTDirect, DTCWTInverse
 from .version import version as __version__
+from .dtcwt.graph import MuReNNDirect
```
[the remark this answers is not in the document]
Comment (xir4n) [reply murenn/__init__.py:17]: ok!

**Review by lostanlen — APPROVED** [review]
Thank you!

**Comment by lostanlen** [comment]
```
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
    unknown package:
        Expected sha256 5ccb288774fdfb07a7e7025ffec286971c06d8d7b4fb162525334616d7629ff9
             Got        d7ab2fa6f65c39cf3ad9138fec65d692f8825a96b349446293db537c5a246c6a
```

python 3.11

**Comment by lostanlen** [comment]
>  I don't know what happened, but after rerunning the workflow, the error disappeared...?

Indeed. Sounds like the problem was unrelated with murenn code and had to do with checksums in the package management system. Regardless, this looks good! Thanks again!
